### PR TITLE
[vcpkg] Add script to remove old archives.

### PR DESCRIPTION
**Pull request by BillyONeal**
Example output:

```
PS D:\vcpkg> .\scripts\azure-pipelines\remove-old-archives.ps1 D:\demo -DaysToKeep 0
Starting remove-old-archives attempt 1 out of 3
Deleted the following:
D:\demo\New folder\deletethis.txt
Failed to delete the following:
D:\demo\New folder\cantdeletethis.txt
Starting remove-old-archives attempt 2 out of 3
Deleted the following:
Failed to delete the following:
D:\demo\New folder\cantdeletethis.txt
Starting remove-old-archives attempt 3 out of 3
Deleted the following:
Failed to delete the following:
D:\demo\New folder\cantdeletethis.txt
```

### CODE REVIEW

**Comment by BillyONeal** [comment]
I can see an argument that we should do this in CI instead. What do folks think?

File: scripts/azure-pipelines/remove-old-archives.ps1 [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: MIT
+#
+
+<#
+.SYNOPSIS
+Deletes files in the indicated archive older than the supplied age.
```
Comment (CaseyCarter) [inline scripts/azure-pipelines/remove-old-archives.ps1:7]: I think you mean the files are "older than the supplied age", not the archive.
```suggestion
Deletes files older than the supplied age in the indicated archive.
```

File: scripts/azure-pipelines/remove-old-archives.ps1 [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#>
+
+[CmdletBinding()]
+Param(
+    [Parameter(Mandatory = $true, Position = 0)][string]$Archive,
+    [int]$DaysToKeep = 7,
+    [int]$Retries = 3
+)
+
+if (Test-Path $Archive) {
```
Comment (CaseyCarter) [inline scripts/azure-pipelines/remove-old-archives.ps1:23]: This branch is much larger than the else branch; reordering and early-return would save a lot of indentation.

File: scripts/azure-pipelines/remove-old-archives.ps1 [hunk 3]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: MIT
+#
+
+<#
+.SYNOPSIS
+Deletes files in the indicated archive older than the supplied age.
+
+.PARAMETER Archive
+Path to the 'archive' directory from which old archives will be deleted.
```
Comment (CaseyCarter) [inline scripts/azure-pipelines/remove-old-archives.ps1:10]: I'm confused: This says we're deleting archives whereas 7 says we're deleting files.

File: scripts/azure-pipelines/remove-old-archives.ps1 [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+<#
+.SYNOPSIS
+Deletes files in the indicated archive older than the supplied age.
+
+.PARAMETER Archive
+Path to the 'archive' directory from which old archives will be deleted.
+
+.PARAMETER DaysToKeep
+Number of days
```
Comment (CaseyCarter) [inline scripts/azure-pipelines/remove-old-archives.ps1:13]: Should we mention that files last accessed more than this many days ago will be selected for deletion?

File: scripts/azure-pipelines/remove-old-archives.ps1 [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    [int]$DaysToKeep = 7,
+    [int]$Retries = 3
+)
+
+if (Test-Path $Archive) {
+    $now = Get-Date
+    $cutoff = $now.AddDays(-$DaysToKeep)
+    for ($thisRetry = 1; $thisRetry -le $Retries; ++$thisRetry) {
+        $failedDeletions = @()
+        Write-Host "Starting remove-old-archives attempt $thisRetry out of $Retries"
```
Comment (CaseyCarter) [inline scripts/azure-pipelines/remove-old-archives.ps1:28]: "out" seems extraneous; "attempt 1 of 3" is perfectly clear.

File: scripts/azure-pipelines/remove-old-archives.ps1 [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+Param(
+    [Parameter(Mandatory = $true, Position = 0)][string]$Archive,
+    [int]$DaysToKeep = 7,
+    [int]$Retries = 3
+)
+
+if (Test-Path $Archive) {
+    $now = Get-Date
+    $cutoff = $now.AddDays(-$DaysToKeep)
+    for ($thisRetry = 1; $thisRetry -le $Retries; ++$thisRetry) {
```
Comment (CaseyCarter) [inline scripts/azure-pipelines/remove-old-archives.ps1:26]: 1-based indexing??!? We've got to you back working on C++ before it's too late! (no change requested)

**Review by CaseyCarter — APPROVED** [review]

File: scripts/azure-pipelines/remove-old-archives.ps1 (hunk 6 above)
Comment (BillyONeal) [reply scripts/azure-pipelines/remove-old-archives.ps1:26]: It's use 1 based indexing or do math when printing the attempt count, so this is the lesser of evils.

**Comment by BillyONeal** [comment]
Unfortunately this doesn't work: it looks like LastAccessTime isn't getting updated when we restore a port with binary caching, so we need some way to exclude hashes that we know of.

File: scripts/azure-pipelines/remove-old-archives.ps1 (hunk 4 above)
Comment (PhoebeHui) [reply scripts/azure-pipelines/remove-old-archives.ps1:13]: How about removing the archives based on port name and version, not all archives?

Comment (BillyONeal) [reply scripts/azure-pipelines/remove-old-archives.ps1:13]: @PhoebeHui The archives are only stored by hash, we don't know what the name and version are. We need a mechanism to enumerate all the possible hashes and not delete those in order to be done with this.

**Comment by BillyONeal** [comment]
Given that this needs near total rework, closing for now.

**Pull request closed without merging**